example/src/main/java/io/netty/example/tunechoserver/TunEchoServer.java

```diff
@@ -0,0 +1,452 @@
+                InetUtils.UDP_HEADER_OFFSET_SOURCE_PORT, udpHeader.destPort);
+        packetBuf.setShort(ipHeader.headerLength +
+                InetUtils.UDP_HEADER_OFFSET_DEST_PORT, udpHeader.sourcePort);
+    }
+
+    private static void changeICMPTypeAndCode(ByteBuf packetBuf, InetUtils.IPHeader ipHeader,
+            InetUtils.ICMPHeader icmpHeader, int newType, int newCode) {
+
+        // Read the portion of the packet containing the ICMP type and code.
+        byte[] oldTypeAndCodeBytes = new byte[2];
+        packetBuf.getBytes(ipHeader.headerLength, oldTypeAndCodeBytes);
+
+        // Read the portion of the packet containing the original ICMP checksum.
+        byte[] oldICMPChecksumBytes = new byte[2];
+        packetBuf.getBytes(ipHeader.headerLength + InetUtils.ICMP_HEADER_OFFSET_CHECKSUM, oldICMPChecksumBytes);
+
+        // Construct a byte array containing the new type and code in the form these fields will appear in
+        // updated the packet.
+        byte[] newTypeAndCode = new byte[2];
+        newTypeAndCode[0] = (byte) newType;
+        newTypeAndCode[1] = (byte) newCode;
+
+        // Update the packet with the new type and code.
+        packetBuf.setBytes(ipHeader.headerLength, newTypeAndCode);
+
+        // Compute a new ICMP checksum based on the updated type and code.
+        int newICMPChecksum = InetUtils.updateInetChecksum(oldTypeAndCodeBytes, 0,
+                newTypeAndCode, 0, 2, icmpHeader.checksum);
+
+        // Update the checksum field in the ICMP header.
+        packetBuf.setShort(ipHeader.headerLength + InetUtils.ICMP_HEADER_OFFSET_CHECKSUM, newICMPChecksum);
+
+        // If the packet is an ICMP packet (vs ICMPv6)...
+        if (ipHeader.ipVersion == InetUtils.IP_VERSION_4) {
+
+            // Read the portion of the packet containing the new ICMP checksum bytes.
+            byte[] newICMPChecksumBytes = new byte[2];
+            packetBuf.getBytes(ipHeader.headerLength + InetUtils.ICMP_HEADER_OFFSET_CHECKSUM, newICMPChecksumBytes);
+
+            // Compute an updated IP checksum.
+            int newIPChecksum = InetUtils.updateInetChecksum(oldICMPChecksumBytes, 0, newICMPChecksumBytes,
+                    0, 2, ipHeader.checksum);
+            newIPChecksum = InetUtils.updateInetChecksum(oldTypeAndCodeBytes, 0, newTypeAndCode, 0, 2, newIPChecksum);
+
+            // Update the checksum field in the IPv4 header.
+            packetBuf.setShort(InetUtils.IPV4_HEADER_OFFSET_CHECKSUM, newIPChecksum);
+        }
+    }
+
+    private static boolean readProps() {
+
+        _tunDeviceName = System.getProperty("tun-device-name", "tun-echo-server");
+
+        try {
+            _threadPoolSize = Integer.parseInt(System.getProperty("thread-pool-size", "0"));
+            if (_threadPoolSize < 0) {
+                throw new IllegalArgumentException();
+            }
+        } catch (Exception ex) {
+            System.err.println("Invalid value specified for thread-pool-size property: " +
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)